src/Discovery/src/Eureka/AppInfo/ApplicationInfoCollection.cs

```diff
@@ -92,79 +97,51 @@ internal void Add(ApplicationInfo app)
         }
     }
 
-    private void AddInstanceToVip(InstanceInfo instance, string address, ConcurrentDictionary<string, ConcurrentDictionary<string, InstanceInfo>> dictionary)
+    private static HashSet<string> ExpandVipAddresses(InstanceInfo instance)
     {
-        lock (_addRemoveInstanceLock)
-        {
-            string addressUpper = address.ToUpperInvariant();
-
-            if (!dictionary.TryGetValue(addressUpper, out ConcurrentDictionary<string, InstanceInfo>? instances))
-            {
-                instances = new ConcurrentDictionary<string, InstanceInfo>();
-                dictionary[addressUpper] = instances;
-            }
+        HashSet<string> addresses = [];
 
-            instances[instance.InstanceId] = instance;
+        if (instance.SecureVipAddress != null)
+        {
+            string[] secureAddresses = instance.SecureVipAddress.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
+            addresses.UnionWith(secureAddresses);
         }
-    }
 
-    private static string[] ExpandVipAddresses(string? addresses)
-    {
-        if (string.IsNullOrWhiteSpace(addresses))
+        if (instance.VipAddress != null)
         {
-            return [];
+            string[] vipAddresses = instance.VipAddress.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
+            addresses.UnionWith(vipAddresses);
         }
 
-        return addresses.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
+        return addresses;
     }
 
-    internal void RemoveInstanceFromVip(InstanceInfo instance)
+    internal void RemoveFromVipInstanceMap(InstanceInfo instance)
     {
         ArgumentNullException.ThrowIfNull(instance);
 
-        foreach (string vipAddress in ExpandVipAddresses(instance.VipAddress))
-        {
-            RemoveInstanceFromVip(instance, vipAddress, VipInstanceMap);
-        }
-
-        foreach (string secureVipAddress in ExpandVipAddresses(instance.SecureVipAddress))
-        {
-            RemoveInstanceFromVip(instance, secureVipAddress, SecureVipInstanceMap);
-        }
-    }
-
-    private void RemoveInstanceFromVip(InstanceInfo instance, string address,
-        ConcurrentDictionary<string, ConcurrentDictionary<string, InstanceInfo>> dictionary)
-    {
-        lock (_addRemoveInstanceLock)
+        foreach (string vipAddress in ExpandVipAddresses(instance))
         {
-            string addressUpper = address.ToUpperInvariant();
+            string addressUpper = vipAddress.ToUpperInvariant();
 
-            if (dictionary.TryGetValue(addressUpper, out ConcurrentDictionary<string, InstanceInfo>? instances))
+            if (VipInstanceMap.TryGetValue(addressUpper, out ConcurrentDictionary<string, InstanceInfo>? instancesById))
             {
-                _ = instances.TryRemove(instance.InstanceId, out _);
-
-                if (instances.IsEmpty)
```

Comment:
It looks like we will end up with zombie apps in the internal dictionary that have no instances ([example in the tests](https://github.com/SteeltoeOSS/Steeltoe/pull/1604/files#diff-180e7a90b742496441a438262e20c27594b7bb4dc89857a53bbbbc10cc928aa8R171)) after this change. Is that intended, inconsequential or am I just missing something?

Reply:
Yes, that's what this text in the PR description is about:

> Locking was removed to improve throughput, which comes at the cost of a small increase in memory usage: when all instances of a service are removed, an empty instance collection is kept in memory.

`GetInstancesByVipAddress` loops over the entries, so it doesn't affect the outcome.